Upgrade to webpack@5

ixrock opened the pull request:
This also fixing nasty caching bug with UI files after combination(s) of `make clean && make dev` / page reload / `git checkout` branches / etc.

**TODO:**

- [x] fix unit tests
- [x] fix integration tests
- [x] fix warnings in compilation of `main` process _(types must be imported separately in files participating in extension-api re-exports)_ 
- [x] fix incorrect import warnings in compilation of `renderer` process
- [x] use `@pmmmwh/react-refresh-webpack-plugin` with `react-refresh-typescript` to support HMR in most cases

ixrock committed:
attempt to upgrade webpack@5 and all relevant packages to latest version -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
debugging webpack@5 usage -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on webpack.renderer.ts:
```diff
@@ -46,7 +57,6 @@ export function webpackLensRenderer({ showVars = true } = {}): webpack.Configura
     },
     output: {
       libraryTarget: "global",
-      library: "",
```
Not supported in `webpack@5`. What is was used for? @jakolehm maybe you know?

ixrock committed:
Merge remote-tracking branch 'origin/master' into webpack5-upgrade-attempt

ixrock committed:
- attempt to use "react-refresh-typescript" with webpack@5 thingy
- extending getTSLoader() to accept more options from ts-loader

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
more fixes, updating webpack loaders

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into webpack5-upgrade-attempt

# Conflicts:
#	package.json
#	yarn.lock

ixrock committed:
merge-fixes, using internal webpack@5 asset handlers (type: "asset/*")

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: raw-loader / <Icon svg="./*">, updated mocked types for importing resources via webpack

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into webpack5-upgrade-attempt

# Conflicts:
#	src/common/catalog-entities/kubernetes-cluster.ts

ixrock committed:
removing webpack-dev-server, clean up

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into webpack5-upgrade-attempt

# Conflicts:
#	package.json
#	yarn.lock

ixrock committed:
fix master-merge conflict

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix/reverted: use sourceMap for styles

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix lint

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: loading svg icons inline as data-url (workaround for "?raw" as it fails in tests and "!!raw-loader!" seems doesn't work at all in webpack@5)

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/common/catalog-entities/kubernetes-cluster.ts:
```diff
@@ -134,7 +135,7 @@ class KubernetesClusterCategory extends CatalogCategory {
   public metadata = {
     name: "Clusters",
-    icon: require("./icons/kubernetes.svg?raw"),
+    icon: KubeClusterCategoryIcon,
```
Woot! 🎉

ixrock replied:
By default imported `*.svg` files contain data-url (string) that can be consumed normally by `<img src>` or with `<Icon svg/>` as inline/raw xml icon.

aleksfront replied:
Amazing!

ixrock committed:
attempt to use webpack-dev-server via node-api -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
attempt to use webpack-dev-server via node-api -- part 2

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: incorrect parsing svg-icon xml-content by <HotbarEntityIcon/>

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: more random fixes related to incorrect parsing svg-icon content, added static Icon.isSvg(content: string)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
attempt to fix: "Uncaught (in promise) DOMException: A network error occurred." when loading cluster frame -- part 1

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
Merge remote-tracking branch 'origin/master' into webpack5-upgrade-attempt

ixrock committed:
fix: loading cluster frame + error in loading terminal default font (which was causing network exception error and hidden cluster-view)
fix: proxying websockets (e.g. terminal)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: wait lensProxy.listen() to obtain lensProxy.port for webpack-dev-server

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix lint

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix unit tests

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/main/index.ts:
```diff
@@ -43,18 +43,26 @@ import { initTray } from "./tray/tray";
-import extensionDiscoveryInjectable from "../extensions/extension-discovery/extension-discovery.injectable";
-import directoryForExesInjectable from "../common/app-paths/directory-for-exes/directory-for-exes.injectable";
-import initIpcMainHandlersInjectable from "./initializers/init-ipc-main-handlers/init-ipc-main-handlers.injectable";
+import lensProtocolRouterMainInjectable
```
Can we not have these all on dual lines?

ixrock replied:
Changes reverted. It was probably auto-formatting files via [.idea](https://github.com/lensapp/lens/tree/master/.idea) settings. Why do we even need to commit them to the repo? @jansav any clue?

Nokel81 commented on src/main/window-manager.ts:
```diff
@@ -34,17 +34,14 @@ export class WindowManager extends Singleton {
   protected disposers: Record<string, Function> = {};
 
   @observable activeClusterId: ClusterId;
+  @observable mainContentUrl = `http://localhost:${LensProxy.getInstance().port}`;
```
I don't think this needs to be `@observable`

ixrock replied:
Removed.

Nokel81 commented on src/renderer/components/dock/terminal/terminal.ts:
```diff
@@ -26,11 +27,14 @@ export class Terminal {
-    const fontPath = require("../../fonts/roboto-mono-nerd.ttf").default; // eslint-disable-line @typescript-eslint/no-var-requires
-    const fontFace = new FontFace(defaultTerminalFontFamily, `url(${fontPath})`);
+    try {
+      const fontFace = new FontFace(defaultTerminalFontFamily, `url(${RobotoMonoFontUrl})`);
```
Is it possible to move this to the `HTML` file?

ixrock replied:
Preloading fonts moved to `template.html` and `fonts.scss`

Nokel81 commented on webpack.main.ts:
```diff
@@ -6,23 +6,23 @@
     target: "electron-main",
-    mode: isProduction ? "production" : "development",
-    devtool: isProduction ? "source-map" : "cheap-eval-source-map",
+    mode: isDevelopment ? "development" : "production",
```
Why was this switched?

ixrock replied:
Because it is the only place where `isProduction` was used when `isDevelopment` used in multiple places.
Also it's done to streamline code with other `webpack` configurations.

Nokel81 replied:
Sounds good, I didn't notice that you also removed `isProduction` as a variable but reducing that set is good.

ixrock committed:
reverted auto-formatted imports with 2 lines (.idea/webstorm)

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: handle warnings in main-process compilation files

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: handle warnings in "renderer" compilation

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: move app fonts preloading into html/css files, clean up

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on src/main/menu/menu.ts:
```diff
@@ -9,13 +9,11 @@ import { appName, isMac, isWindows, docsUrl, supportUrl, productName } from "../
 import { checkForUpdates, isAutoUpdateEnabled } from "../app-updater";
 import type { MenuRegistration } from "./menu-registration";
 
-export type MenuTopId = "mac" | "file" | "edit" | "view" | "help";
```
Not used anywhere.

Nokel81 replied:
It should be, for the registration of menu items since we don't support adding new menu items.

ixrock replied:
Well, it was used before. Dunno who was refactored menu that it now has some leftovers. 
Anyway, this is out of scope.

Nokel81 replied:
I think it is out of scope to remove it.

ixrock replied:
Nope. This is just like removing a comment since it's not used / doesn't touch any other code.

Nokel81 replied:
Disagree, an upgrade PR like this should be only doing the minimal work to get it working, this should be a separate PR.

ixrock replied:
Okay, I'll revert it, but just want to say.. With that frenetic attitude in PR's you probably gonna work alone on this project ~(maybe this is your goal)~.. Just see how other projects has quick PR reviews without PITA..

ixrock committed:
update "webpack@5", "typescript@4.5" in bundled extensions / attempt to fix weird errors in build pipeline:

ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/common/catalog/catalog-category-registry.d.ts
6:27-35
[tsl] ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/common/catalog/catalog-category-registry.d.ts(6,28)
      TS1005: ',' expected.

ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/renderer/components/select/select.d.ts
8:14-28
[tsl] ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/renderer/components/select/select.d.ts(8,15)
      TS1005: ',' expected.

ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/renderer/api/catalog-entity-registry.d.ts
8:14-22
[tsl] ERROR in /home/runner/work/lens/lens/src/extensions/npm/extensions/dist/src/renderer/api/catalog-entity-registry.d.ts(8,15)
      TS1005: ',' expected.

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: handle errors in bundled extensions compilation process

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: "webpack" not found in production

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: removed preloading fonts in template via <link preload> since it's not bundled with HtmlWebpackPlugin() anyway anda all fonts loaded via css @font-face rule

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
apply HMR at least for css/styles and use manual page reload on app/scripts change

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
use `react-refresh-typescript` and `@pmmmwh/react-refresh-webpack-plugin` to support HMR in most cases

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on extensions/kube-object-event-status/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "kube-object-event-status",
-  "version": "0.0.1",
+  "version": "5.4.0-beta.1.1644584997435",
```
Can you revert this change?

ixrock replied:
Reverted.

Nokel81 commented on extensions/kube-object-event-status/package.json:
```diff
@@ -18,8 +18,8 @@
-    "ts-loader": "^8.0.4",
-    "typescript": "^4.3.2",
-    "webpack": "^4.46.0"
+    "ts-loader": "latest",
```
We shouldn't use `latest` here because it will break if there is a mismatch of versions between this file and `../../package.json`

ixrock replied:
Why it will break when it's not used in "production" / inside packed extension?
I think only `ts-loader` must be compatible with local `webpack` package.

Nokel81 replied:
We do run `npm install` on extensions, even bundled extensions.

ixrock replied:
When we do? So maybe we should move it to `peerDependencies` then?
Anyway I think it should be handled from parent's package version, otherwise it's just copy-pasting all around. Can we use actual versions from app's `package.json` on extensions install?

Nokel81 replied:
https://github.com/lensapp/lens/blob/master/src/extensions/extension-discovery/extension-discovery.ts#L383-L406

This is where we do the install calls

ixrock replied:
Well, it's still unclear how this installing dependency handle unpacking extension:

```ts
await this.dependencies.installExtension(extension.absolutePath);
```

In general it feels wrong if compiled extension require matching own development dependencies with the  app (in case of `react`, `mobx` and some other packages we're sharing them via global env because version matching there is important for app's work. @jakolehm @nevalla @panuhorsmalahti wdyt?

jakolehm replied:
@Nokel81 @ixrock Lens installs extension npm package without development dependencies -> extension dev dependencies don't need to match same versions on Lens side.

jakolehm replied:
Note: using `latest` version might still be a bit risky (in general).

ixrock replied:
I think it's not more risky than just having outdated specific version == in both cases I assume it's easy to spot/fix/use what is needed per extension package.

Nokel81 commented on src/renderer/components/+network-ingresses/ingress-details.tsx:
```diff
@@ -15,7 +15,7 @@ import { ResourceMetrics } from "../resource-metrics";
 import { IngressCharts } from "./ingress-charts";
 import { KubeObjectMeta } from "../kube-object-meta";
-import { getBackendServiceNamePort, getMetricsForIngress, IIngressMetrics } from "../../../common/k8s-api/endpoints/ingress.api";
+import { getBackendServiceNamePort, getMetricsForIngress, type IIngressMetrics } from "../../../common/k8s-api/endpoints/ingress.api";
```
Wondering out loud, is this down automatically? (Because it would be great if it was)

ixrock replied:
No idea.. but those warnings just in compilation process and probably could be ignored.

Nokel81 commented on webpack.extensions.ts:
```diff
@@ -6,9 +6,11 @@
+import { cssModulesWebpackRule, fontsLoaderWebpackRules, iconsAndImagesWebpackRules } from "./webpack.renderer";
 
 export default function generateExtensionTypes(): webpack.Configuration {
+  const { isDevelopment, isProduction } = vars;
```
Shouldn't we switch to just using `isDevelopment` here too?

ixrock replied:
Yes, why not.

ixrock committed:
responding to comments

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
revered extension version in package-lock.json (autoupdated on `make dev`)

Signed-off-by: Roman <ixrock@gmail.com>

jakolehm reviewed (approved):
LGTM

Nokel81 reviewed (approved):
Except for the font issue. This looks good to me.
